### PR TITLE
Export to Spreadsheet Text Formatting

### PR DESCRIPTION
**Pull request by nabalone**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)

### CODE REVIEW

File: src/BloomExe/Spreadsheet/SpreadsheetIO.cs [hunk 1]
```diff
@@ -80,9 +111,6 @@ public static void WriteSpreadsheet(InternalSpreadsheet spreadsheet, string outp
 						}
 					}
 				}
-				// Review: is this helpful? Excel typically makes very small cells, so almost
```
Comment (nabalone) [inline src/BloomExe/Spreadsheet/SpreadsheetIO.cs:83]: Do we still want this comment now that we can adjust column widths? Not wrapping the text makes the cells run over each other or get cut off, which I personally find to be less intuitive

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> Reviewed 9 of 10 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 7 unresolved discussions (waiting on @nabalone)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 13 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnPKdz4awK4OdJ78st:-MfnPKdz4awK4OdJ78su:b-r294wo) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/MarkedUpText.cs#L13)):*
> ```Smalltalk
> 	/// of it bolded, italicized, underlined, and/or superscripted
> 	/// </summary>
> 	public class MarkedUpText : List<MarkedUpTextRun>
> ```

It bothers me a bit that MarkedUpText inherits publicly from List. For one thing, it's giving away an implementation secret. By inheriting from List, MarkedUpText has all kinds of functions it doesn't need, like Sort(). For another, it automatically makes the class mutable...and where possible, immutable classes are nice to have for all kinds of reasons. It's not much of a problem for a non-library class that probably won't be widely used, but I wonder how hard it would be to make the List a member variable?
(If you go with another suggestion of moving ParseXml into this class, that method would still have full access to the list.)
Unless we also give the class a function to write itself to an Excel Cell...tempting but less obviously a natural responsibility of the class...you'd need a function to get at the runs. But it could be as simple as IEnumerable<MarkedUpTextRun> Runs => m_list, which lets a client perform a foreach over the runs, but doesn't allow it to modify the list.
___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 15 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnHlXAAPxtLka9ThXT:-MfnHlXAAPxtLka9ThXU:bszwxd0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/MarkedUpText.cs#L15)):*
> ```Smalltalk
> 	public class MarkedUpText : List<MarkedUpTextRun>
> 	{
> 		public override string ToString()
> ```

This doesn't need doing for this PR, but when you do import, it may be useful to give this class a method that produces the XML representation of the string with its formatting. That might be an even more useful ToString()...though it's worth keeping a 'just the text' method too.
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 76 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnJTYZ0611zjwa5Tbw:-MfnJTYZ0611zjwa5Tbx:bb0jijz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L76)):*
> ```Smalltalk
> 							MarkedUpText markedUpText = ParseXml(content);
> 							ExcelRange currentCell = worksheet.Cells[r, c + 1];
> 							currentCell.IsRichText = true;
> ```

It's likely that upwards of 95% of Bloom text blocks don't have any internal formatting, in which case, it would be fine to just set the Value. I'm not sure how much overhead there is to making a cell RichText, so maybe I'm being guilty of premature optimization...but if you go with the suggestion below of always creating the MarkedUpText object, it might be as simple as adding "&& markedUpText.HasFormatting" to the IF (plus implementing that property).
(One of the possible virtues of an XML spreadsheet import/export capability is that other people could, just possibly, write their own code to manipulate the spreadsheet, whether in DotNet, or as spreadsheet macros, or using a language like XSLT that is designed for manipulating XML. Such code might be simpler if the author knows he doesn't use formatting and can therefore skip dealing with rich text. So that would be another reason not to do RichText unless we need it.)

(I also find myself wondering whether "WriteToExcelCell" would be a nice method to add to MarkedUpText. It's tricky...this class wants to be responsible for Excel IO, MarkedUpText wants to be responsible for working with formatted text, this function is both. So you can make a good case either way. What do you think?)
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 94 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnK3FM2oVWDxks2ugW:-MfnK3FM2oVWDxks2ugX:b-x70mq2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L94)):*
> ```Smalltalk
> 						else
> 						{
> 							worksheet.Cells[r, c + 1].Value = content;
> ```

I think here we've lost the logic handling paragraph elements, from the old SpreadsheetCell.GetContent() method (and which you have reimplmeneted in ParseXml).
One way to get it back might be to always create the MarkedUpText, but if retainMarkup is false, just get the text from it.
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 168 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnM7po1PWcfiZ41L0m:-MfnM7po1PWcfiZ41L0n:b1dizl0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L168)):*
> ```Smalltalk
> 		/// Adds newlines after paragraphs, but drops leading and trailing, but not intermediate, white space.
> 		/// </summary>
> 		public static MarkedUpText ParseXml(string xmlString)
> ```

I'm thinking that this method (and its helpers) would fit nicely in the MarkedUpText class itself. It could well be useful if we ever have another client for the class.
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 182 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnXVph9y57KLDSAM-K:-MfnXVph9y57KLDSAM-L:b-e2blof) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L182)):*
> ```Smalltalk
> 				//remove leading and trailing whitespace. We don't want a trailing newline which
> 				//will make excel put a blank line at the end of the cell
> 				while (markedUpText.Count > 0 && string.IsNullOrWhiteSpace(markedUpText[0].Text))
> ```

I think there are cases where we want to keep leading white space. For example, suppose the first thing is an empty paragraph? Or what if there are a few spaces followed immediately by a bold word?
___
*[src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs, line 89 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnTQ4Q5IjmSccYuur2:-MfnTQ4Q5IjmSccYuur3:bye7qho) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs#L89)):*
> ```Smalltalk
> 			var result = xmlresult.ToString();
> 			Assert.That(result, Is.EqualTo(expected));
> 		}
> ```

Probably worth adding one more test case, either here or as its own method, to check that it handles an empty string as expected.

**Review by nabalone — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 7 unresolved discussions (waiting on @JohnThomson and @nabalone)

___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 76 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnJTYZ0611zjwa5Tbw:-MfoBhXs8RdXcgZK1WsB:bhrpx2m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L76)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It's likely that upwards of 95% of Bloom text blocks don't have any internal formatting, in which case, it would be fine to just set the Value. I'm not sure how much overhead there is to making a cell RichText, so maybe I'm being guilty of premature optimization...but if you go with the suggestion below of always creating the MarkedUpText object, it might be as simple as adding "&& markedUpText.HasFormatting" to the IF (plus implementing that property).
(One of the possible virtues of an XML spreadsheet import/export capability is that other people could, just possibly, write their own code to manipulate the spreadsheet, whether in DotNet, or as spreadsheet macros, or using a language like XSLT that is designed for manipulating XML. Such code might be simpler if the author knows he doesn't use formatting and can therefore skip dealing with rich text. So that would be another reason not to do RichText unless we need it.)

(I also find myself wondering whether "WriteToExcelCell" would be a nice method to add to MarkedUpText. It's tricky...this class wants to be responsible for Excel IO, MarkedUpText wants to be responsible for working with formatted text, this function is both. So you can make a good case either way. What do you think?)
</blockquote></details>

Are you thinking of HasFormatting as a parameter that gets passed in/"manually" set at some point, or as something that the MarkedUpText objects detects and keeps track of?

It makes more sense in my brain personally for the IO class to be doing all of the input and output

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 7 unresolved discussions (waiting on @nabalone)

___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 76 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnJTYZ0611zjwa5Tbw:-MfoHzYf0bmtq8WX_TtO:b-pn1si0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L76)):*
<details><summary><i>Previously, nabalone wrote…</i></summary><blockquote>

Are you thinking of HasFormatting as a parameter that gets passed in/"manually" set at some point, or as something that the MarkedUpText objects detects and keeps track of?

It makes more sense in my brain personally for the IO class to be doing all of the input and output
</blockquote></details>

Sorry, I was thinking of HasFormatting as a property of MarkedUpText: HasFormatting => Runs.Any(r => r.HasFormatting), and then run would have HasFormatting => Bold | Italic |Underlined|Superscript.

It's fine to leave the code that writes to the Excel Cell in the IO class.

**Review by nabalone — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: 6 of 10 files reviewed, 6 unresolved discussions (waiting on @JohnThomson and @nabalone)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 13 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnPKdz4awK4OdJ78st:-Mg6pprz0sz0Nrdpz4Lo:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/MarkedUpText.cs#L13)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It bothers me a bit that MarkedUpText inherits publicly from List. For one thing, it's giving away an implementation secret. By inheriting from List, MarkedUpText has all kinds of functions it doesn't need, like Sort(). For another, it automatically makes the class mutable...and where possible, immutable classes are nice to have for all kinds of reasons. It's not much of a problem for a non-library class that probably won't be widely used, but I wonder how hard it would be to make the List a member variable?
(If you go with another suggestion of moving ParseXml into this class, that method would still have full access to the list.)
Unless we also give the class a function to write itself to an Excel Cell...tempting but less obviously a natural responsibility of the class...you'd need a function to get at the runs. But it could be as simple as IEnumerable<MarkedUpTextRun> Runs => m_list, which lets a client perform a foreach over the runs, but doesn't allow it to modify the list.
</blockquote></details>

Done.
___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 15 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnHlXAAPxtLka9ThXT:-Mg6puHU2U92aIVfNiq_:b32w1t) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/MarkedUpText.cs#L15)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This doesn't need doing for this PR, but when you do import, it may be useful to give this class a method that produces the XML representation of the string with its formatting. That might be an even more useful ToString()...though it's worth keeping a 'just the text' method too.
</blockquote></details>

Ok
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 94 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnK3FM2oVWDxks2ugW:-Mg737MzDWOqFs4va90E:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L94)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think here we've lost the logic handling paragraph elements, from the old SpreadsheetCell.GetContent() method (and which you have reimplmeneted in ParseXml).
One way to get it back might be to always create the MarkedUpText, but if retainMarkup is false, just get the text from it.
</blockquote></details>

Done.
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 168 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnM7po1PWcfiZ41L0m:-Mg7393B-1oIeLU5NPD1:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L168)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I'm thinking that this method (and its helpers) would fit nicely in the MarkedUpText class itself. It could well be useful if we ever have another client for the class.
</blockquote></details>

Done.
___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 182 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnXVph9y57KLDSAM-K:-MfoO21K6LA_0aRSnztT:b-m2swie) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L182)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think there are cases where we want to keep leading white space. For example, suppose the first thing is an empty paragraph? Or what if there are a few spaces followed immediately by a bold word?
</blockquote></details>

Ok, I made it keep all leading white space and updated the tests accordingly.
___
*[src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs, line 89 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnTQ4Q5IjmSccYuur2:-Mg73Wrn9sqdV0yTdjnG:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs#L89)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Probably worth adding one more test case, either here or as its own method, to check that it handles an empty string as expected.
</blockquote></details>

Done.

**Review by nabalone — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: 5 of 10 files reviewed, 6 unresolved discussions (waiting on @JohnThomson)

___
*[src/BloomExe/Spreadsheet/SpreadsheetIO.cs, line 76 at r2](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MfnJTYZ0611zjwa5Tbw:-Mg74du88RHO0w8J0iVR:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b8f4da74c5b4aaf328935ff0087efa0ed47e037b/src/BloomExe/Spreadsheet/SpreadsheetIO.cs#L76)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Sorry, I was thinking of HasFormatting as a property of MarkedUpText: HasFormatting => Runs.Any(r => r.HasFormatting), and then run would have HasFormatting => Bold | Italic |Underlined|Superscript.

It's fine to leave the code that writes to the Excel Cell in the IO class.
</blockquote></details>

Done.

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> Reviewed 3 of 5 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 3 unresolved discussions (waiting on @nabalone)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 24 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7WjCp4eZiyZkSeJxN:-Mg7WjCq1zdmmCEeXmSp:b-k8nvsc) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomExe/Spreadsheet/MarkedUpText.cs#L24)):*
> ```Smalltalk
> 		}
> 
> 		public string PlainText()
> ```

I like the idea of overriding ToString(). Could just return PlainText() for now...by the time you get import working, there might be a better option that also indicates the markup, maybe as XHTML
___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 84 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7bAlBDiKhRysOVsxl:-Mg7bAlBDiKhRysOVsxm:b-h3w0s2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomExe/Spreadsheet/MarkedUpText.cs#L84)):*
> ```Smalltalk
> 					continue;
> 				}
> 				if (x.Name == "p")
> ```

As this stands, it will discard completely (not even put in pending) anything that is not <p> and isn't whitespace. I think I may have told you that was OK, but seeing it in code made me think again and take another look at what I had before.
I think we want something more like my original, which adds the pending elements and then the content for ANY element that gets this far (that is, other than all-white-space elements that aren't <p>). Then, IF it's a <p>, put a \r\n run in pending. In other words, I think only the last line of this block should be conditional.
(Sorry again, I think that contradicts what I told you verbally. I think it will be quite unusual, maybe never happen, to find a mixture of <p>s and other things at the root of a bloom-editable. But if we do, I'd hate to just discard anything EXCEPT white space.)
It's worth adding a couple of test cases, too, that non-white stuff is not lost, whether at the start, between paragraphs, or after the last paragraph, and whether it's at the top level or, say, inside a <div> instead of a <p>.
We won't be able to get such blocks back unchanged when re-importing, but at least text won't get lost completely.
___
*[src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs, line 82 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7__8D8fcfJ86O3OrJ:-Mg7__8D8fcfJ86O3OrK:b-bjpkoo) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs#L82)):*
> ```Smalltalk
> 		[TestCase(@"<p>Some text.<span class=""bloom-linebreak""></span>Some more.</p>", "Some text.\r\nSome more.")]
> 		[TestCase("\r\n\t\tText without p tags.\r\n\tSome more.\r\n", "\r\n\t\tText without p tags.\r\n\tSome more.\r\n")]
> 		public void ParsesFormattedXmlHandlesWhitespace(string input, string expected)
> ```

If I'm reading the code right, white space following the last </p> is intended to be dropped...but text would not be?

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 3 unresolved discussions (waiting on @nabalone)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 84 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7bAlBDiKhRysOVsxl:-Mg7frwSEqCO0_0U6y_P:b-34fee) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomExe/Spreadsheet/MarkedUpText.cs#L84)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

As this stands, it will discard completely (not even put in pending) anything that is not <p> and isn't whitespace. I think I may have told you that was OK, but seeing it in code made me think again and take another look at what I had before.
I think we want something more like my original, which adds the pending elements and then the content for ANY element that gets this far (that is, other than all-white-space elements that aren't <p>). Then, IF it's a <p>, put a \r\n run in pending. In other words, I think only the last line of this block should be conditional.
(Sorry again, I think that contradicts what I told you verbally. I think it will be quite unusual, maybe never happen, to find a mixture of <p>s and other things at the root of a bloom-editable. But if we do, I'd hate to just discard anything EXCEPT white space.)
It's worth adding a couple of test cases, too, that non-white stuff is not lost, whether at the start, between paragraphs, or after the last paragraph, and whether it's at the top level or, say, inside a <div> instead of a <p>.
We won't be able to get such blocks back unchanged when re-importing, but at least text won't get lost completely.
</blockquote></details>

Sorry, that last comment was totally trashed wherever I used text that looked like HTML markers. Here we go again. Read wedges for square brackets.
As this stands, it will discard completely (not even put in pending) anything that is not [p] and isn't whitespace. I think I may have told you that was OK, but seeing it in code made me think again and take another look at what I had before.
I think we want something more like my original, which adds the pending elements and then the content for ANY element that gets this far (that is, other than all-white-space elements that aren't [p]s). Then, IF it's a [p], put a \r\n run in pending. In other words, I think only the last line of this block should be conditional.
(Sorry again, I think that contradicts what I told you verbally. I think it will be quite unusual, maybe never happen, to find a mixture of [p]s and other things at the root of a bloom-editable. But if we do, I'd hate to just discard anything EXCEPT white space.)
It's worth adding a couple of test cases, too, that non-white stuff is not lost, whether at the start, between paragraphs, or after the last paragraph, and whether it's at the top level or, say, inside a [div] instead of a [p].
We won't be able to get such blocks back unchanged when re-importing, but at least text won't get lost completely.

**Review by nabalone — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: 8 of 10 files reviewed, 3 unresolved discussions (waiting on @JohnThomson)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 24 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7WjCp4eZiyZkSeJxN:-Mg7lH7z7sFfaxILoaF7:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomExe/Spreadsheet/MarkedUpText.cs#L24)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I like the idea of overriding ToString(). Could just return PlainText() for now...by the time you get import working, there might be a better option that also indicates the markup, maybe as XHTML
</blockquote></details>

Done.
___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 84 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7bAlBDiKhRysOVsxl:-Mg7qQNsEWH5WwmNjjAO:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomExe/Spreadsheet/MarkedUpText.cs#L84)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Sorry, that last comment was totally trashed wherever I used text that looked like HTML markers. Here we go again. Read wedges for square brackets.
As this stands, it will discard completely (not even put in pending) anything that is not [p] and isn't whitespace. I think I may have told you that was OK, but seeing it in code made me think again and take another look at what I had before.
I think we want something more like my original, which adds the pending elements and then the content for ANY element that gets this far (that is, other than all-white-space elements that aren't [p]s). Then, IF it's a [p], put a \r\n run in pending. In other words, I think only the last line of this block should be conditional.
(Sorry again, I think that contradicts what I told you verbally. I think it will be quite unusual, maybe never happen, to find a mixture of [p]s and other things at the root of a bloom-editable. But if we do, I'd hate to just discard anything EXCEPT white space.)
It's worth adding a couple of test cases, too, that non-white stuff is not lost, whether at the start, between paragraphs, or after the last paragraph, and whether it's at the top level or, say, inside a [div] instead of a [p].
We won't be able to get such blocks back unchanged when re-importing, but at least text won't get lost completely.
</blockquote></details>

Done.
___
*[src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs, line 82 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7__8D8fcfJ86O3OrJ:-Mg7rDVj3_z8hsq4yfc_:b-1cpayq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs#L82)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

If I'm reading the code right, white space following the last </p> is intended to be dropped...but text would not be?
</blockquote></details>

Is this better?

**Review by JohnThomson — APPROVED** [review]
> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622)* status: all files reviewed, 3 unresolved discussions (waiting on @JohnThomson and @nabalone)

___
*[src/BloomExe/Spreadsheet/MarkedUpText.cs, line 36 at r6](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-MgBI9G10GSIxi57K9Sr:-MgBI9G10GSIxi57K9Ss:b-hbmuoq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2e368ef9f8a15bc3e693543d1ba08c2f860f1992/src/BloomExe/Spreadsheet/MarkedUpText.cs#L36)):*
> ```Smalltalk
> 		//TODO test this method
> 		//TODO handle newlines/paragraphs
> 		public override string ToString()
> ```

When you do import, we'll probably want to rename this method something like ToXhtml(). Then, make ToString() (which is mainly for the debugger) simply return ToXhtml(). That allows other code to use a method that's not dependent on what we decide we'd like to see in the debugger. But this is fine for now, when no other code is using it.
___
*[src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs, line 82 at r5](https://reviewable.io/reviews/bloombooks/bloomdesktop/4622#-Mg7__8D8fcfJ86O3OrJ:-MgBHu6t9Yvyj6CPp7wM:b-sp2oyq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/7d4f442ca6d57f23f3e107a876a0f0955f05b8a0/src/BloomTests/Spreadsheet/SpreadsheetFormattingTests.cs#L82)):*
<details><summary><i>Previously, nabalone wrote…</i></summary><blockquote>

Is this better?
</blockquote></details>

I think so. We might not know for sure until we find some real documents with these oddities.
